### DISCUSSION TITLE
Much better minterm optimizier

### DISCUSSION THREAD

**Discussion by gh228df** [logisim-evolution/logisim-evolution#2020] [General]
Hi everyone,

I've tried building my custom CPU and have been frustrated by the slow speed of minterm optimization in Logisim. It struggles to optimize functions with 14 inputs and consumes a significant amount of RAM.

To address this, I developed and implemented my own optimization algorithm. You can find it in my repository: https://github.com/gh228df/boolean-optimizer

My algorithm is 800 times more memory efficient and approximately 100 times faster for smaller functions. For functions with higher input counts, the performance improvement is even greater. It can also be easily parallelized.

However sometimes it produces longer functions just by a couple of minterms, but I have also seen it produce a shorter functions, so I think it is pretty much the same. For shorter functions there is no difference at all. Also the output function is much shorter when logisim combines minterms.

I hope this helps you.

**Comment by maehne** [comment]
Thanks for your message and offer for an improved Boolean optimization algorithm. Your repository is scarce on details on what work it is based and how it compares to established optimisation algorithms such as the ESPRESSO heuristic logic minimizer. Regarding your comparison with Logisim-evolution, did you try also with one of the latest [nightly builds](https://github.com/logisim-evolution/logisim-evolution/actions/runs/10379041181)?

As Logisim-evolution is primarilly a teaching tool, we are not too much concerned about performance problems for such big logic functions. In many cases, the problems can be avoided by partitioning the logic functions into smaller sub-functions, which are then assembled in a structural circuit often combined with multiplexers.

**Reply by gh228df** [reply]
Hello maehne, 

The algorithm my program is based on was crafted by me, that's why the information in the repo is so shallow. My target was to use as many bitwise and simple boolean operations as possible to get the fastest runtime. I have combined simple boolean algebra with half-function-bruteforce and something like Quine–McCluskey algorithm to get it. I think for small functions it doesn't matter how fast it runs. I can totally agree with you on that. I am quite surprised that it gets so close to logisim's algorithm while running so much faster, that's why I posted the original message. Maybe some of the optimizations can be incorporated with your algorithm to get starting amount of minterms significantly down.

**Reply by maehne** [reply]
Thanks for the explanation. I think that deserves to be documented. If you want that your work is reused by others, you should consider adding proper file headers and clearly stating the licence of the source code under which you would like it distributed.

By the way, @BFH-ktt1 started an effort in PR #1530 to integrate the Espresso algorithm into Logisim-evolution. Unfortunately, he lacked time to finish the work.
